src/main/scala/org/neo4j/spark/reader/Neo4jInputPartitionReader.scala

```diff
@@ -0,0 +1,75 @@
+package org.neo4j.spark.reader
+
+import java.time.{LocalDate, LocalDateTime, OffsetTime, ZoneOffset, ZonedDateTime}
+
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.util.{ArrayData, DateTimeUtils}
+import org.apache.spark.sql.sources.v2.reader.{InputPartition, InputPartitionReader}
+import org.apache.spark.sql.types.StructType
+import org.apache.spark.unsafe.types.UTF8String
+import org.neo4j.driver.internal.{InternalPoint2D, InternalPoint3D}
+import org.neo4j.driver.{Record, Session}
+import org.neo4j.spark.{DriverCache, Neo4jOptions, Neo4jQuery, QueryType}
+import java.sql.Timestamp
+
+import org.neo4j.spark.util.Neo4jUtil
+
+import collection.JavaConverters._
+
+class Neo4jInputPartitionReader(private val options: Neo4jOptions,
+                                private val schema: StructType,
+                                private val jobId: String) extends InputPartition[InternalRow] with InputPartitionReader[InternalRow] {
+
+  var customIterator: Iterator[Record] = _
+  var session: Session = _
+  var driverCache: DriverCache = new DriverCache(options.connection, jobId)
+
+  override def createPartitionReader(): InputPartitionReader[InternalRow] = new Neo4jInputPartitionReader(options, schema, jobId)
+
+  def next: Boolean = {
+    if (customIterator == null) {
+      session = driverCache.getOrCreate().session()
+      // @todo use readTransaction
+      customIterator = session.run(Neo4jQuery.build(options.query)).list.asScala.iterator
+    }
+
+    customIterator.hasNext
+  }
```

Comment:
la gestione della transaction la dobbiamo fare subito altrimenti ce lo cassano:
```suggestion
  def next: Boolean = {
    if (result == null) {
      session = driverCache.getOrCreate().session(options.session.toNeo4jSession()) // crea il metodo
      transaction = session.beginTransaction()
      result = transaction.run(Neo4jQuery.build(options.query))
    }
    result.hasNext
  }
```

Reply:
@utnaf ricordami che dobbiamo parlare dell'accessMode lo dobbiamo mettere subito poi ti spiego perché